cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1215,24 +1215,60 @@ struct get_page_schema {
   __device__ size_type operator()(gpu::PageInfo const& page) { return page.src_col_schema; }
 };
 
+struct input_col_info {
+  int const schema_idx;
+  size_type const nesting_depth;
+};
+
+/**
+ * @brief Converts a 1-dimensional index into page, depth and column indices used in
+ * allocate_columns to compute columns sizes.
+ *
+ * The input index will iterate through pages, nesting depth and column indices in that order.
+ */
+struct reduction_indices {
+  const size_t _page_idx;
+  const size_type _depth_idx;
+  const size_type _col_idx;
+
+  __device__ reduction_indices(size_t index, size_type max_depth, size_t num_pages)
+    : _page_idx(index % num_pages),
+      _depth_idx((index / num_pages) % max_depth),
+      _col_idx(index / (max_depth * num_pages))
```

Comment:
Since the member variables are public, do not use underscore prefix for them. To differentiate between member vars and parameters which have similar name, add underscore suffix to the parameter. For example:
```suggestion
  __device__ reduction_indices(size_t index_, size_type max_depth_, size_t num_pages_)
    : page_idx(index_ % num_pages_),
      depth_idx((index_ / num_pages_) % max_depth_),
      col_idx(index_ / (max_depth_ * num_pages_))
```

Reply:
In addition, this constructor can be removed entirely because the struct instance can be initialized by initializer list.

Reply:
I don't think that's a good idea.  Doing that puts the burden of understanding how we are mapping things on the caller.  The purpose of this struct is to hide that.

Reply:
Yeah that sounds reasonable!